helios-services/src/main/java/com/spotify/helios/master/reaper/GceAgentReaper.java

```diff
@@ -0,0 +1,155 @@
+    this(masterModel, pubsub, topic, timeoutHours, SYSTEM_CLOCK, PERMITS_PER_SECOND,
+         new Random().nextInt(DELAY));
+  }
+
+  @VisibleForTesting
+  GceAgentReaper(final MasterModel masterModel,
+                 final PubSub pubsub,
+                 final String topic,
+                 final long timeoutHours,
+                 final Clock clock,
+                 final double permitsPerSecond,
+                 final int initialDelay) {
+    super(permitsPerSecond, initialDelay, DELAY, TIME_UNIT);
+    this.masterModel = masterModel;
+    this.pubsub = pubsub;
+    this.topic = topic;
+    checkArgument(timeoutHours > 0);
+    this.timeoutMillis = TimeUnit.HOURS.toMillis(timeoutHours);
+    this.clock = clock;
+  }
+
+  @Override
+  protected void startUp() throws Exception {
+    super.startUp();
+
+    if (pubsub.getSubscription(SUBSCRIPTION) == null) {
+      final SubscriptionInfo subscriptionInfo = SubscriptionInfo.newBuilder(topic, SUBSCRIPTION)
+          // put the message back in the queue if it's not consumed within one iteration
+          .setAckDeadLineSeconds((int) TIME_UNIT.convert(DELAY, TimeUnit.SECONDS))
+          .build();
+      pubsub.create(subscriptionInfo);
+    }
+  }
+
+  @Override
+  Iterable<ReceivedMessage> collectItems() {
+    // only pull as many messages as we can process in one iteration
+    final int maxMessages = (int) (TIME_UNIT.convert(DELAY, TimeUnit.SECONDS) / PERMITS_PER_SECOND);
+    return ImmutableList.copyOf(pubsub.pull(SUBSCRIPTION, maxMessages));
+  }
+
+  @Override
+  void processItem(final ReceivedMessage message) {
+    try {
+      // we ack before processing. something could still go wrong, but it's preferable to miss
+      // a few deregistrations that should then be caught by DeadAgentReaper rather than block
+      // the queue due to exceptions
+      message.ack();
+
+      final LogEntry logEntry = LogEntry.parseFrom(message.getPayload().asInputStream());
+      final Struct payload = logEntry.getJsonPayload();
+
+      if (!"GCE_OPERATION_DONE".equals(payload.getFieldsOrThrow("event_type").getStringValue())) {
+        return;
+      }
+      if (!"compute.instances.delete".equals(payload.getFieldsOrThrow("event_subtype")
+                                                 .getStringValue())) {
+        return;
+      }
+
+      final String host = logEntry.getResource().getLabelsOrThrow("name");
+
+      if (masterModel.isHostUp(host)) {
+        // Host UP -- nothing to do
+        return;
+      }
+
+      final AgentInfo agentInfo = masterModel.getAgentInfo(host);
+      if (agentInfo == null) {
+        return;
+      }
+
+      try {
+        log.info("Reaping deleted GCE instance '{}'");
+        masterModel.deregisterHost(host);
+      } catch (Exception e) {
+        log.warn("Failed to reap deleted GCE instance '{}'", host, e);
+      }
+    } catch (Exception e) {
+      log.warn("Failed to parse pubsub message: {}", message, e);
```

Comment:
Also, iirc Message.toString() logs a message JD and just a portion of the message as hex, not the full message contents, in case you are expecting that to be logged.